### DISCUSSION TITLE
Android系统怎么实现protectSocket调用

### DISCUSSION THREAD

**Discussion by qxhdev** [eycorsican/leaf#403] [Q&A]
/// Example:
///
/// Note: the protect method could be identified as unused code on the Android
/// side and stripped by ProGuard, you may need to add a keep rule to let
/// ProGuard knows we need the code, e.g.:
///
/// -keep class com.leaf.and.aleaf.** { *; }
///
/// // Sets a callback method to protect sockets.
/// //
/// // Expects a method with the given name and signature `(I)Z`.
/// #[allow(non_snake_case)]
/// #[no_mangle]
/// pub unsafe extern "system" fn Java_com_leaf_and_aleaf_SimpleVpnService_setProtectSocketCallback(

看了文档 但是还是不知道怎么实现

**Comment by eycorsican** [comment]
```kotlin
class SimpleVpnService: VpnService() {

...

    private external fun setProtectSocketCallback(name: String)
    
    ...
    
    
    fun protectSocket(fd: Int): Boolean {
        return protect(fd)
    }
    
    ...
    
            setProtectSocketCallback("protectSocket")
            
```

**Reply by qxhdev** [reply]
谢谢回复！！ 我后面研究了半天调通了，大概步骤是这样:
1. 先要在rust源码里加入回调代码
#[cfg(target_os = "android")]
#[allow(non_snake_case)]
#[no_mangle]
pub unsafe extern "system" fn Java_com_xxx_xxxn_vpn_LeafJni_setCallback(
    mut env: JNIEnv,
    _object: JObject,
    class: JClass,
    name: JString,
) {
    let Ok(name) = env.get_string(&name) else {
        return;
    };
    let name: String = name.into();
    if let Ok(class_g) = env.new_global_ref(class) {
        leaf::mobile::callback::android::set_protect_socket_callback(class_g, name);
    }
}

#[cfg(target_os = "android")]
#[allow(non_snake_case)]
#[no_mangle]
pub unsafe extern "system" fn JNI_OnLoad(vm: JavaVM, _: *mut std::os::raw::c_void) -> jint {
    leaf::mobile::callback::android::set_jvm(vm);
    JNI_VERSION_1_6
}

#[cfg(target_os = "android")]
#[allow(non_snake_case)]
#[no_mangle]
pub unsafe extern "system" fn JNI_OnUnload(_vm: JavaVM, _: *mut std::os::raw::c_void) {
    leaf::mobile::callback::android::unset_protect_socket_callback();
    leaf::mobile::callback::android::unset_jvm();
}
2. 重新编译Aandroid库
3. 在VpnService里通过jni调用

调用LeafJni.setCallback时传入 自己的VpnService实现对象 和 protect方法名
   我是这样调用:
  override fun onCreate() {
        leafJni.setCallback(this, "protect")
    }

**Reply by 52yaoer** [reply]
你好，可以分享下这个可以编译的工程么？我自己设置的老是不对，

**Comment by 52yaoer** [comment]
我这里引入报这个错误

leaf::mobile::callback::android::set_protect_socket_callback(class_g, name);
   |         ------------------------------------------------------------ ^^^^^^^ expected `GlobalRef`, found a different `GlobalRef`

leaf::mobile::callback::android::set_jvm(vm);
    |     ---------------------------------------- ^^ expected `JavaVM`, found a different `JavaVM`
